base/src/main/java/io/spine/base/Environment.java

```diff
@@ -21,37 +21,76 @@
+import static io.spine.base.BaseEnvironmentType.PRODUCTION;
+import static io.spine.base.BaseEnvironmentType.TESTS;
+
 /**
  * Provides information about the environment (current platform used, etc.).
+ *
+ * <p>By default, knows only about {@link BaseEnvironmentType}. Library users may extend the list
+ * of known environment types by providing their environment {@code enum} to
+ * {@linkplain #registerCustom(Class)}.
+ *
+ * <p><b>When extending, please note</b> that this class does not handle the situations when two
+ * or more {@linkplain EnvironmentType environment types} return {@code true} on the
+ * {@link EnvironmentType#currentlyOn()}. As such, if two or more user-defined environment types
+ * think that they are currently on, <b>the behaviour of {@link #envType()} is undefined.</b>
  */
 @SPI
-@SuppressWarnings("AccessOfSystemProperties") // OK as we need system properties for this class.
 public final class Environment {
 
-    private static final Environment INSTANCE = new Environment();
+    private static final ImmutableList<EnvironmentType> BASE_TYPES =
+            ImmutableList.of(TESTS, PRODUCTION);
 
-    /**
-     * The key name of the system property which tells if a code runs under a testing framework.
-     *
-     * <p>If your testing framework is not among the supported by {@link #isTests()},
-     * set this property to {@code true} before running tests.
-     */
-    public static final String ENV_KEY_TESTS = "io.spine.tests";
+    private static final Environment INSTANCE = new Environment();
 
-    /** If set, tells if the code runs from a testing framework. */
-    private @Nullable Boolean tests;
+    private ImmutableList<EnvironmentType> knownEnvTypes;
+    private @Nullable EnvironmentType currentEnvType;
 
-    /** Prevents instantiation of this singleton class from outside. */
     private Environment() {
+        this.knownEnvTypes = BASE_TYPES;
     }
 
     /** Creates a new instance with the copy of the state of the passed environment. */
     private Environment(Environment copy) {
-        this.tests = copy.tests;
+        this.knownEnvTypes = copy.knownEnvTypes;
+        this.currentEnvType = copy.currentEnvType;
+    }
+
+    /**
+     * Registers all of the constants from the provided {@code EnvironmentType}-implementing
+     * {@code enum}.
+     *
+     * <p>If the specified {@code enum} has already been registered, throws
+     * an {@code IllegalStateException}.
+     *
+     * <p>Note that the {@linkplain BaseEnvironmentType default types} are still present.
+     * When trying to {@linkplain #envType() determine which environment type} is currently on,
+     * the user defined types are checked first.
+     *
+     * @param enumClass
+     *         an enum class that specifies the environment types
+     * @param <E>
+     *         a type that defines possible environment types
+     */
+    public static <E extends Enum & EnvironmentType> void registerCustom(Class<E> enumClass) {
```

Comment:
As discussed in person, we may get rid of the explicit registration and make our peace with it.